rfcs/0112-demote-darwin-support-tier.md

```diff
@@ -0,0 +1,91 @@
+---
+feature: demote_darwin_support_tier
+start-date: 2021-10-27
+author: piegames
+co-authors: many
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+Move x86_64-darwin from a Tier 2 platform to Tier 3, as described in [RFC 0046](https://github.com/NixOS/rfcs/blob/master/rfcs/0046-platform-support-tiers.md).
+
+# Motivation
+[motivation]: #motivation
+
+Darwin is constantly adding additional maintainer burden. Especially, it does not live up to the requirements:
+
+- "A lot of packages built by Hydra, full ofBorg support."
+    - There is only one ofBorg builder. It can take several hours or days for an ofBorg build to finish for Darwin, if it finishes at all.
+    - Hydra builders a greatly understaffed.
+    - If Apple follows through on [their announcement to discontinue](https://www.businessinsider.com/apple-macbook-pro-discontinued) x86-based devices, the situation won't improve due to a physical lack of hardware.
+- "Most packages work, credible ambition to reach Tier 1 coverage at some point."
+    - Tier 1 support is far out of reach.
```

Comment:
Do we have a number we need to hit here? The RFC uses `most`, `many`, `almost everything`, and those adjectives match my personal experience: I haven't had any issue with getting the packages I needed on darwin, sometimes the package was marked as broken on darwin but the fix was very simple, sometimes it was marked as linux only but it worked without any changes on darwin.

If we have a specific % of packages that is needed to address this, we might be able to achieve it by going for those low hanging fruits.

Reply:
> Do we have a number we need to hit here?

Again, we haven't. You can consider this a flaw in RFC 46. More generally though, Support Tier 1 is much more than simply the number of supported packages. Again, the phrasing of RFC 46 is misleading: these should have been two separate independent points.

> sometimes it was marked as linux only but it worked without any changes on darwin

Then the package is a testimony of Darwin having been broken at one point in time in its history.

Reply:
> Then the package is a testimony of Darwin having been broken at one point in time in its history.

This is probably an overstatement, hard to be sure whether the expression author said «looks like it could be Linux-specific, it's not that I can cheaply check if it _works_ on Darwin, whatever»

Reply:
As far as I can tell, many `nixpkgs` developers don't care about Darwin to the point of not even thinking about explicitly disabling Darwin support. I consider the following far more likely: "CI fails for Darwin and I don't see an easy way to fix this, whatever".

Reply:
Well, as Nixpkgs is consistently bad at using tristate «yes/no/don't know» instead of boolean, I would expect just as many people say «so, where I _know_ it works? Linux? let's put that». Knowing the nomenclature beyond Linux requires care already… 

(It's usually about `platforms`, not `broken`, so you list positive claims there)

Reply:
> sometimes the package was marked as broken on darwin but the fix was very simple, sometimes it was marked as linux only but it worked without any changes on darwin

I meant to specify both cases:

1. `isBroken` was true on darwin, that definitely points to it having been broken on darwin at some point.
2. `platforms` was set to linux, probably because whoever added the package didn't test it on any other platform.

I'm definitely guilty of the same thing, I set `platforms` to `platforms.linux ++ platforms.darwin` when I add new packages, even though it might support other platforms as well, I just don't know how and don't want to invest the effort in testing them.

Reply:
There are also many transitive broken on darwin packages. So packages where eval fails because they have a nixpkgs dependency on eg. systemd or other software that doesn't support darwin even though the package itself can compile just fine on darwin if that dependency was removed for darwin.